components/storage/blockdevice/COMPONENT_SDIO/SDIOBlockDevice.cpp

```diff
@@ -0,0 +1,393 @@
+#define SD_BLOCK_DEVICE_ERROR_UNSUPPORTED_BLOCKSIZE -5012 /*!< unsupported blocksize, only 512 byte supported */
+#define SD_BLOCK_DEVICE_ERROR_READBLOCKS -5013            /*!< read data blocks from SD failed */
+#define SD_BLOCK_DEVICE_ERROR_WRITEBLOCKS -5014           /*!< write data blocks to SD failed */
+#define SD_BLOCK_DEVICE_ERROR_ERASEBLOCKS -5015           /*!< erase data blocks to SD failed */
+
+#define BLOCK_SIZE_HC 512 /*!< Block size supported for SD card is 512 bytes  */
+
+// Types
+#define SDCARD_NONE 0  /**< No card is present */
+#define SDCARD_V1 1    /**< v1.x Standard Capacity */
+#define SDCARD_V2 2    /**< v2.x Standard capacity SD card */
+#define SDCARD_V2HC 3  /**< v2.x High capacity SD card */
+#define CARD_UNKNOWN 4 /**< Unknown or unsupported card */
+
+SDIOBlockDevice::SDIOBlockDevice(PinName cardDetect) : _cardDetect(cardDetect),
+    _is_initialized(0),
+    _sectors(0),
+    _init_ref_count(0)
+{
+    // Only HC block size is supported.
+    _block_size = BLOCK_SIZE_HC;
+    _erase_size = BLOCK_SIZE_HC;
+}
+
+SDIOBlockDevice::~SDIOBlockDevice()
+{
+    if (_is_initialized) {
+        deinit();
+    }
+}
+
+int SDIOBlockDevice::init()
+{
+    debug_if(SD_DBG, "init Card...\r\n");
+
+    lock();
+
+    if (!_is_initialized) {
+        _init_ref_count = 0;
+    }
+
+    _init_ref_count++;
+
+    if (_init_ref_count != 1) {
+        unlock();
+        return BD_ERROR_OK;
+    }
+
+    if (isPresent() == false) {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_DEVICE;
+    }
+
+    int status = sdio_init();
+    if (BD_ERROR_OK != status) {
+        unlock();
+        return BD_ERROR_DEVICE_ERROR;
+    }
+
+    sdio_get_card_info(&_cardInfo);
+    _is_initialized = true;
+    debug_if(SD_DBG, "SD initialized: type: %d  version: %d  class: %d\n",
+             _cardInfo.CardType, _cardInfo.CardVersion, _cardInfo.Class);
+    debug_if(SD_DBG, "SD size: %d MB\n",
+             _cardInfo.LogBlockNbr / 2 / 1024);
+
+    // get sectors count from cardinfo
+    _sectors = _cardInfo.LogBlockNbr;
+    if (BLOCK_SIZE_HC != _cardInfo.BlockSize) {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_UNSUPPORTED_BLOCKSIZE;
+    }
+
+    unlock();
+    return status;
+}
+
+int SDIOBlockDevice::deinit()
+{
+    debug_if(SD_DBG, "deinit Card...\r\n");
+    lock();
+
+    if (!_is_initialized) {
+        _init_ref_count = 0;
+        unlock();
+        return BD_ERROR_OK;
+    }
+
+    _init_ref_count--;
+
+    if (_init_ref_count) {
+        unlock();
+        return BD_ERROR_OK;
+    }
+
+    int status = sdio_deinit();
+    _is_initialized = false;
+
+    _sectors = 0;
+
+    unlock();
+    return status;
+}
+
+int SDIOBlockDevice::read(void *buffer, bd_addr_t addr, bd_size_t size)
+{
+    int status = 0;
+    lock();
+    if (isPresent() == false) {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_DEVICE;
+    }
+
+    if (!is_valid_read(addr, size)) {
```

Comment:
when not initialized 'this->size()" is 0 and is_valid_read/program/trim will fail.. 
You will receive SD_BLOCK_DEVICE_ERROR_PARAMETER instead of SD_BLOCK_DEVICE_ERROR_NO_INIT, which is more appropriate.